### PR TITLE
Position estimation

### PR DESCRIPTION
**Pull request by icai0**
use 4 cameras to estimate robot position with apriltags

### CODE REVIEW

File: constants.py [hunk 1]
```diff
@@ -217,11 +217,34 @@
     Pose3d(0.236, 0.206, 0.197, Rotation3d()),
 )
 
-kLimelightPoseKey = "vision/limelight"
+kEstimatedPoseKey = "vision/estimatedPose"
```
Comment (lmaxwell24) [inline constants.py]: whats the purpose of this constant? I don't see it used anywhere

File: constants.py [hunk 2]
```diff
@@ -217,11 +217,34 @@
...
 kPhotonvisionCameraName = "camcam"
+kPhotonvisionCameraArray = ["frontLeft", "frontRight", "backLeft", "backRight"]
 
+kPhotonvisionFrontLeftCameraKey = "frontLeftCamera"
+kPhotonvisionFrontRightCameraKey = "frontRightCamera"
+kPhotonvisionBackLeftCameraKey = "backLeftCamera"
+kPhotonvisionBackRightCameraKey = "backRightCamera"
+
+
+kRobotToFrontLeftCameraTransform = Transform3d(
```
Comment (lmaxwell24) [inline constants.py:230]: fine for now but ideally each of these is a different transform

File: subsystems/visionsubsystem.py [hunk 3]
```diff
@@ -1,76 +1,88 @@
...
+
+            self.updateAdvantagescopePose(Pose3d() + bestRelativeTransform, camera)
 
-        botPose = (
-            Pose3d()
-            + bestRelativeTransform
-            + constants.kLimelightRelativeToRobotTransform.inverse()
-        )
+            cameraTransforms = {
+                "frontLeft": constants.kRobotToFrontLeftCameraTransform,
```
Comment (lmaxwell24) [inline subsystems/visionsubsystem.py]: this, as well as the keys, and the cameras should get grouped into a larger struct like EstimatedPose
something like the following
```py
class VisionCamera:
    transform: Transform3d
    camera: PhotonCamera
    poseKey: str
```

File: subsystems/visionsubsystem.py [hunk 4]
```diff
@@ -1,76 +1,88 @@
...
-        limelightPose3d = (
-            pose3dFrom2d(self.estimatedPosition)
-            + constants.kLimelightRelativeToRobotTransform
-        )
-        limelightPose = advantagescopeconvert.convertToSendablePoses([limelightPose3d])
+    def updateAdvantagescopePose(
+        self, cameraPose3d: Pose3d, camera: PhotonCamera
+    ) -> None:
+        cameraKeys = {
+            "frontLeft": constants.kPhotonvisionFrontLeftCameraKey,
```
Comment (lmaxwell24) [inline subsystems/visionsubsystem.py]: see above comment with regards to merging this into a larger struct, purpose being to allow a more dynamic number of cameras

**Review by lmaxwell24 — CHANGES_REQUESTED** [review]

**Review by lmaxwell24 — APPROVED** [review]

**Review by LandonBayer — APPROVED** [review]

**Pull request merged**